loader/db.go

```diff
@@ -34,213 +36,151 @@ import (
-	return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, strings.Join(sqls, ";"))
+	return conn.executeSQLCustomRetry(ctx, sqls, retry.IsLoaderRetryableError)
 }
 
-func executeSQLImp(ctx *tcontext.Context, db *sql.DB, sqls []string) error {
-	var (
-		err error
-		txn *sql.Tx
-		res sql.Result
-	)
-
-	txn, err = db.Begin()
-	if err != nil {
-		ctx.L().Error("fail to begin a transaction", zap.String("sqls", fmt.Sprintf("%-.200v", sqls)), zap.Error(err))
-		return terror.DBErrorAdapt(err, terror.ErrDBDriverError)
+func (conn *Conn) executeSQLCustomRetry(ctx *tcontext.Context, sqls []baseconn.SQL, retryFn func(err error) bool) error {
+	if conn == nil || conn.baseConn == nil {
+		return terror.ErrDBUnExpect.Generate("database connection not valid")
 	}
 
-	for i := range sqls {
-		ctx.L().Debug("execute statement", zap.String("sqls", fmt.Sprintf("%-.200v", sqls[i])))
-		res, err = txn.ExecContext(ctx.Context(), sqls[i])
-		if err != nil {
-			ctx.L().Warn("execute statement", zap.String("sqls", fmt.Sprintf("%-.200v", sqls[i])), log.ShortError(err))
-			rerr := txn.Rollback()
-			if rerr != nil {
-				ctx.L().Error("fail rollback", log.ShortError(rerr))
-			}
-			return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, sqls[i])
-		}
-		// check update checkpoint successful or not
-		if i == 2 {
-			row, err1 := res.RowsAffected()
-			if err1 != nil {
-				ctx.L().Warn("fail to get rows affected", zap.String("sqls", fmt.Sprintf("%-.200v", sqls[i])), log.ShortError(err1))
-				continue
+	params := retry.Params{
+		RetryCount:         10,
+		FirstRetryDuration: 2 * time.Second,
+		BackoffStrategy:    retry.LinearIncrease,
+		IsRetryableFn: func(retryTime int, err error) bool {
+			ctx.L().Debug("execute statement", zap.Int("retry", retryTime), zap.String("sqls", fmt.Sprintf("%-.200v", sqls)))
+			tidbExecutionErrorCounter.WithLabelValues(conn.cfg.Name).Inc()
+			if retryFn(err) {
+				return true
 			}
-			if row != 1 {
-				ctx.L().Warn("update checkpoint", zap.Int64("affected rows", row))
+			return false
+		},
+	}
+	_, _, err := conn.baseConn.RetryStrategy.Apply(
+		ctx,
+		params,
+		func(ctx *tcontext.Context) (interface{}, error) {
+			startTime := time.Now()
+			_, err := conn.baseConn.ExecuteSQL(ctx, sqls)
+			failpoint.Inject("LoadExecCreateTableFailed", func(val failpoint.Value) {
+				errCode, err1 := strconv.ParseUint(val.(string), 10, 16)
+				if err1 != nil {
+					ctx.L().Fatal("failpoint LoadExecCreateTableFailed's value is invalid", zap.String("val", val.(string)))
+				}
+
+				if len(sqls) == 1 && strings.Contains(sqls[0].Query, "CREATE TABLE") {
+					err = tmysql.NewErr(uint16(errCode))
+					ctx.L().Warn("executeSQLCustomRetry failed", zap.String("failpoint", "LoadExecCreateTableFailed"), zap.Error(err))
+				}
+			})
+			cost := time.Since(startTime)
+			txnHistogram.WithLabelValues(conn.cfg.Name).Observe(cost.Seconds())
+			if cost > 1 {
+				ctx.L().Warn("transaction execute successfully", zap.Duration("cost time", cost))
```

Comment:
always `successfully`? whether needing to check `err`?

Reply:
actually I didn't get what this mean, I just copy from old code

Reply:
@3pointer but old code check `err` first, see https://github.com/pingcap/dm/blob/cb56783af3dfbee6682b6e45147078fe4bd5d5c5/loader/db.go#L161

Reply:
oh I see, I fixed it in https://github.com/pingcap/dm/pull/236/files#diff-5f5bb90e4d87dad0cb25c377bb10679bR132